lib/katello/tasks/update_subscription_facet_backend_data.rake

```diff
@@ -0,0 +1,61 @@
+namespace :katello do
+  task :update_subscription_facet_backend_data => ["environment"] do
+    def error(message)
+      @errors << message
+      @errors << "\n"
+    end
+
+    def report_errors
+      if @errors.any?
+        filename = "subscription_facet_upgrade-#{Time.now.to_i}.log"
+        path = "/var/log/foreman/#{filename}"
+        path = "/tmp/#{filename}" unless File.writable?(path)
+
+        file = File.open(path, 'w')
+        @errors.each { |error| file.write(error) }
+        file.close
+        $stderr.print "***********************************\n"
+        $stderr.print "*************WARNING***************\n"
+        $stderr.print "Errors detected during upgrade step.\n"
+        $stderr.print "Details saved to: #{file.path}\n"
+        $stderr.print "This step can be rerun with:\n"
+        $stderr.print "  foreman-rake katello:update_subscription_facet_backend_data\n"
+        $stderr.print "You are likely encountering a bug.\n"
+        $stderr.print "***********************************\n"
+      end
+    end
+
+    @errors ||= []
+    User.current = User.anonymous_api_admin
+    puts _("Updating backend data for subscription facets")
+
+    #there may be some invalid hosts, if there are create a primary interface
+    ::Host.includes(:interfaces).find_each do |host|
+      if host.primary_interface.nil?
+        host.interfaces.create!(:primary => true)
```

Comment:
Also a lot of these hacky workaround are due to the fact that hosts can be created or edited and put into an invalid state.  We probably should have just put 'save(:validate => false)' everywhere we needed to make sure a host got updated but so far we've tended to avoid that.  Not being able to count on the host object to valid has consistently been a very difficult issue.